### ISSUE TITLE
Unable to create more than 4 networks.

### ISSUE THREAD

**Issue by jeffutter** [docker/libnetwork#1101]
## Issue

It seems as though I am unable to create more than 4 networks before I start getting `Error response from daemon: failed to parse pool request for address space "LocalDefault" pool "" subpool "": could not find an available predefined network` errors.
## Example

Active Networks:

```
# docker network ls
NETWORK ID          NAME                                       DRIVER
5849372c9920        bridge                                     bridge
4aed776fb7db        host                                       host
93eba83fc4bd        workspace_default                          bridge
b16948421d33        none                                       null
```

Which is the three defaults and one created by docker-compose.

Then when I attempt to create additional networks the following happens.

```
# for i in $(seq 1 40); do echo $i; docker network create "test$i"; done
1
05cb052f58404eba674d8fd4f9e3e39897c8167966fd3d18b6984789e57bcb81
2
7e6bf138743070f26a17ba285c2d4994350fcb90f59ab748826f2847d2f1c831
3
a3c1e54fc972f1ceac917aafc48a16e5b2231c715ca3285daf3d21c087dd15e7
4
Error response from daemon: failed to parse pool request for address space "LocalDefault" pool "" subpool "": could not find an available predefined network
5
Error response from daemon: failed to parse pool request for address space "LocalDefault" pool "" subpool "": could not find an available predefined network
6
...
```

Creation of all other networks fails until I remove the ones I just created.

It is my understanding that the default pool should allow for ~31 networks. However I'm reaching a total of 7.
## Additional Information

```
# docker --version
Docker version 1.10.2, build c3959b1

# uname -a
Linux jenkins02 3.19.0-49-generic #55~14.04.1-Ubuntu SMP Fri Jan 22 11:24:31 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux

# lsb_release -a
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 14.04.3 LTS
Release:    14.04
Codename:   trusty

# ip route
default via 10.10.9.1 dev eth0
10.10.9.0/24 dev eth0  proto kernel  scope link  src 10.10.9.40
172.17.0.0/16 dev docker0  proto kernel  scope link  src 172.17.0.1
172.19.0.0/16 dev br-61a0b1f644db  proto kernel  scope link  src 172.19.0.1
172.22.0.0/16 dev br-93eba83fc4bd  proto kernel  scope link  src 172.22.0.1
172.99.19.0/24 dev br-0a56d7229a9d  proto kernel  scope link  src 172.99.19.1
172.99.51.0/24 dev br-e6440eac3257  proto kernel  scope link  src 172.99.51.1
192.168.16.0/20 dev br-fe0745b2b6b9  proto kernel  scope link  src 192.168.16.1
```

I suspect this issue is related to: https://github.com/docker/libnetwork/issues/951 and https://github.com/docker/compose/issues/2279

Can anyone provide some insight or suggestions as to how to debug what is going on here?

Thanks

**Comment by aboch** [comment]
@jeffutter 
From your current installed routes, you should be able to create 31 bridge networks.
Can you please post the content of your host's `/etc/resolv/conf` file. Code takes that into account as well when checking for route overlap.

If that does not give us an answer, then we have to look whether some stale configs are left in the local store.

**Comment by jeffutter** [comment]
/etc/resolv.conf:

```
# Dynamic resolv.conf(5) file for glibc resolver(3) generated by resolvconf(8)
#     DO NOT EDIT THIS FILE BY HAND -- YOUR CHANGES WILL BE OVERWRITTEN
nameserver 8.8.8.8
search our.domain.com
```

**Comment by aboch** [comment]
Thanks @jeffutter, no custom nameserver which could overlap with the `172.[16-31].x.x/16` and `192.168.[0-240].x.x/20` predefined pools available for local networks. 
Then it must be some stale network holding the subnet configuration.
Need to look at the `/var/lib/docker/network/files/local-kv.db` file, if you could attach it after removing all your created networks.

Ungraceful shutdown of daemon can lead to this situation. We added code in 1.11 for daemon to cleanup and release resources allocated for stale networks at reboot.

Thanks

**Comment by aboch** [comment]
@jeffutter Once you have a copy of file, as a workaround (also to prove the theory above), you can delete that file and reload the daemon, and see if you are now able to create the max number of bridge networks w/o specifying a subnet during their creation.

**Comment by jeffutter** [comment]
@aboch Here is that file.
[local-kv.db.zip](https://github.com/docker/libnetwork/files/213504/local-kv.db.zip)

After removing that file I was in fact able to create 27 (+ 3 defaults = 30?) networks. I'm not aware that we have had any ungraceful shutdowns of the daemon, but it is possible someone else did that.

I was suspicious that something wasn't being cleared all the time since we are using this on a CI server and it spins up quite a few containers at a time, many many times a day.

I am curious to know what the contents of that db turns up.

In the meantime I'll let it run for a while and see if any more problems crop up.

Thanks.

**Comment by aboch** [comment]
@jeffutter 
I can also confirm the problem is related to the messed local store file.
Looking at its content (`strings local-kv.db`) I can see there is a bunch of stale subnet in the ipam config (under the key `docker/network/v1.0/ipam/default/config/LocalDefault/`)

I am not sure how it ended in that state though, because there is no stale network in the file holding to those.
So even with the new logic of releasing the resources allocated to stale networks on daemon boot your issue would not be fixed. It would be interesting to know how the system ended in that situation, this is the first time we see this.

So, your only workaround is to erase the local store file and reboot.

**Comment by jeffutter** [comment]
@aboch I have run into this same problem again. This time I could manually create no networks. I can confirm that the daemon had not been restarted _at all_.

Here is the current DB before I cleared it 
[local-kv.db.zip](https://github.com/docker/libnetwork/files/219614/local-kv.db.zip)

I have one suspicion as to what might be leading to this:

We are using docker via docker-compose via Jenkins (no plugin, just executing docker-compose commands directly). What I am wondering is if it is possible when a Jenkins job is aborted (Which I believe sends a SIG TERM and then a SIG KILL). If it might not be cleaning up the DB properly? Is there any possibility that this is causing the buildup of networks in the database?

**Comment by dantswain** [comment]
I think I just ran into this same issue.  Very similar setup to @jeffutter - using docker & docker-compose with Jenkins.  Had to delete `network/files/locak-kv.db` and *reboot* the box - daemon restart was not sufficient.  Docker 1.10.3 on CentOS.

**Comment by dantswain** [comment]
@jeffutter Have you found any workarounds for this or had any ideas to try?  I'd be glad to try some things on our setup.  It's getting really frustrating having to reboot our build server so often.

**Comment by aboch** [comment]
@jeffutter @dantswain Although we did not get to how exactly your db ended in this inconsistent state, docker 1.12 will have a change which should also fix your specific issue (#1130). The out of sync between your bridge networks and the respective ipam database should no longer occur.

**Comment by dantswain** [comment]
@aboch Awesome. Any idea when that will be out?

**Comment by aboch** [comment]
@dantswain You can see the docker/docker release progress at https://github.com/docker/docker/releases. 
Current date is: `1.12.0 (2016-07-14)`

**Comment by dantswain** [comment]
@aboch Thanks.  Doesn't look like 1.12 is in the yum repo yet :/

**Comment by dantswain** [comment]
Ah, nevermind, I found it in the testing repo as it's still a release candidate.

I've updated to 1.12.0 rc2 on our build server - will update here if we have any luck (or fail to).

**Comment by aboch** [comment]
@dantswain That's cool you can update your server to use the release candidate.
Please keep us posted. Thanks.

**Comment by dantswain** [comment]
@aboch @jeffutter  Mixed news.  We are still getting this error on 1.12.0 rc2.  HOWEVER it no longer requires a reboot to resolve.  There were around 30 networks 'left over' from Jenkins builds and doing a `docker network rm -$(docker network ls -q)` resolved the error.  Not sure why the networks aren't being cleaned up from this particular build, though.  We're using the same basic workflow for all of our builds and it seems to just be this one.  We call `docker-compose stop` and `docker-compose rm -f` in a `finally` block in our Jenkinsfile.  I'll update here if I find anything else.

**Comment by saada** [comment]
@dantswain , I was running in to the same issue. I found this helpful: https://docs.docker.com/compose/reference/down/

`docker-compose down` which is equivalent of a `stop`, `rm`, and cleans up networks for you.

**Comment by aboch** [comment]
@dantswain Actually that is a good news: You can see the networks in `docker network ls`, and they are ~30 which is the max number of autoallocated bridge networks.
You can remove them and release their ipam resources.

The problem originally reported was that the networks were removed from the list, but the ipam resources associated to some of them were left over in the DB. And could not be reclaimed by newly created networks.

For your other question ,please note the networks are persisted across daemon reload. You always need to explicitely remove them.

Thanks.

**Comment by dantswain** [comment]
@saada That worked like a charm! Thanks! :+1:

@aboch Yea, the key was updating to 1.12.  I'm just trying to leave a trail so others can avoid the same frustrations :)

To recap:
1) Update to docker 1.12
2) Use `docker-compose down` to clean up, not `docker-compose stop` followed by `docker-compose rm -f`

**Comment by codekoala** [comment]
I encountered the ``could not find an available predefined network`` earlier today using docker-compose. None of the tips listed here seemed to help, but creating a network manually did the trick for me:

    docker network create --subnet=172.21.0.0/16 project_default

where ``project_default`` is what ``docker-compose`` said it was trying to use. I wasn't using Jenkins or anything, so the manual command was feasible since I knew ``172.21.0.0/16`` was available.

**Comment by aboch** [comment]
@codekoala 
That could be a symptom of ip route(s) on your system masking all the predefined pools from which docker picks for the networks.
Typical case is a vpn route which captures all the traffic.
Manually specifying a `--subnet` during network creation is known to fix this, as docker will bypass the subnet overlap check on user explicit chosen subnet.

See #779

**Comment by aboch** [comment]
I am going ahead and will close this issue as fixed by #1130, given we have not seen reports of this on docker 1.12.x.

If any new report comes, feel free to add and will reopen.
